In [2]:
# Loading Data
from src.data_preparation import load_data

In [3]:
PROCESSED_DATA_PATH = "data/processed/df_1h_lag_BLV_spatial_images.csv"
df = load_data(PROCESSED_DATA_PATH, date_col="measurement_time")

PROCESSED_DATA_PATH_2 = "data/processed/df_3h_lag_BLV_spatial_images.csv"
df2 = load_data(PROCESSED_DATA_PATH_2, date_col="measurement_time")

Loaded 11,560 records
Date range: 2014-01-03 14:00:00+00:00 to 2016-12-30 23:00:00+00:00
Timezone: UTC
Loaded 10,510 records
Date range: 2014-01-03 14:00:00+00:00 to 2016-12-30 23:00:00+00:00
Timezone: UTC


In [4]:
df.drop(columns=["Unnamed: 0", "timestamp"], inplace=True)

In [5]:
df2.drop(columns=["Unnamed: 0", "timestamp"], inplace=True)

In [6]:
df.head()

ghi         dni              PST_Time_meas  \
measurement_time                                                               
2014-01-03 14:00:00+00:00    0.000000    0.000000  2014-01-03 06:29:30-08:00   
2014-01-03 15:00:00+00:00   12.914500   89.245833  2014-01-03 07:29:30-08:00   
2014-01-03 16:00:00+00:00  143.775500  636.475000  2014-01-03 08:29:30-08:00   
2014-01-03 17:00:00+00:00  289.955000  783.193333  2014-01-03 09:29:30-08:00   
2014-01-03 18:00:00+00:00  402.643333  843.050000  2014-01-03 10:29:30-08:00   

                           solar_zenith  time_gap_hours  time_gap_norm  \
measurement_time                                                         
2014-01-03 14:00:00+00:00    100.276046        0.166667       0.006944   
2014-01-03 15:00:00+00:00     89.724162        0.016667       0.000694   
2014-01-03 16:00:00+00:00     80.146724        0.016667       0.000694   
2014-01-03 17:00:00+00:00     71.988814        0.016667       0.000694   
2014-01-03 18:00:00+00:00     65.816158        0.016667       0.000694   

                           day_boundary_flag  hour_progression  absolute_hour  \
measurement_time                                                                
2014-01-03 14:00:00+00:00           0.016667          6.491667      24.491667   
2014-01-03 15:00:00+00:00           0.000000          7.491667      25.491667   
2014-01-03 16:00:00+00:00           0.000000          8.491667      26.491667   
2014-01-03 17:00:00+00:00           0.000000          9.491667      27.491667   
2014-01-03 18:00:00+00:00           0.000000         10.491667      28.491667   

                               GHI_cs  ...    ENT(G)      AVG(B)     STD(B)  \
measurement_time                       ...                                    
2014-01-03 14:00:00+00:00    0.000000  ...       NaN         NaN        NaN   
2014-01-03 15:00:00+00:00    0.000000  ...  5.564803  151.246458  28.520968   
2014-01-03 16:00:00+00:00   35.159500  ...  5.510188  141.713212  33.481258   
2014-01-03 17:00:00+00:00  184.554711  ...  5.478535  137.213417  35.403827   
2014-01-03 18:00:00+00:00  321.992462  ...  5.456043  138.616210  36.162608   

                             ENT(B)   AVG(RB)   STD(RB)   ENT(RB)  AVG(NRB)  \
measurement_time                                                              
2014-01-03 14:00:00+00:00       NaN       NaN       NaN       NaN       NaN   
2014-01-03 15:00:00+00:00  5.360367  0.808373  0.144332  4.688985 -0.113270   
2014-01-03 16:00:00+00:00  5.414777  0.909988  0.148683  4.742913 -0.053787   
2014-01-03 17:00:00+00:00  5.377718  0.965367  0.142668  4.686905 -0.022647   
2014-01-03 18:00:00+00:00  5.356960  0.947050  0.140950  4.682247 -0.032250   

                           STD(NRB)  ENT(NRB)  
measurement_time                               
2014-01-03 14:00:00+00:00       NaN       NaN  
2014-01-03 15:00:00+00:00  0.086132  4.008980  
2014-01-03 16:00:00+00:00  0.079303  3.913008  
2014-01-03 17:00:00+00:00  0.071275  3.777387  
2014-01-03 18:00:00+00:00  0.072482  3.788302  

[5 rows x 83 columns]

In [7]:
df.columns

Index(['ghi', 'dni', 'PST_Time_meas', 'solar_zenith', 'time_gap_hours',
       'time_gap_norm', 'day_boundary_flag', 'hour_progression',
       'absolute_hour', 'GHI_cs', 'DNI_cs', 'CSI_ghi', 'CSI_dni',
       'season_flag', 'hour_sin', 'hour_cos', 'month_sin', 'month_cos',
       'nam_ghi', 'nam_dni', 'nam_cc', 'PST_Time_nam', 'nam_target_time',
       'B_CSI_ghi_8h', 'V_CSI_ghi_8h', 'L_CSI_ghi_8h', 'B_CSI_ghi_9h',
       'V_CSI_ghi_9h', 'L_CSI_ghi_9h', 'B_CSI_ghi_10h', 'V_CSI_ghi_10h',
       'L_CSI_ghi_10h', 'B_CSI_ghi_11h', 'V_CSI_ghi_11h', 'L_CSI_ghi_11h',
       'B_CSI_ghi_12h', 'V_CSI_ghi_12h', 'L_CSI_ghi_12h', 'B_CSI_ghi_13h',
       'V_CSI_ghi_13h', 'L_CSI_ghi_13h', 'B_CSI_ghi_14h', 'V_CSI_ghi_14h',
       'L_CSI_ghi_14h', 'B_CSI_ghi_15h', 'V_CSI_ghi_15h', 'L_CSI_ghi_15h',
       'B_CSI_ghi_16h', 'V_CSI_ghi_16h', 'L_CSI_ghi_16h', 'B_CSI_ghi_17h',
       'V_CSI_ghi_17h', 'L_CSI_ghi_17h', 'B_CSI_ghi_18h', 'V_CSI_ghi_18h',
       'L_CSI_ghi_18h', 'B_CSI_ghi_19h', 'V_CSI_ghi_19h',

In [8]:
# 2. Rolling Origin Split (More Suitable in TimeSeries) like K-Folds
from src.data_preparation import rolling_origin_evaluation,save_splits_info

rollingSplits_df = rolling_origin_evaluation(df=df, start_train = '2014-01-2',
    end_train = '2016-12-31')
save_splits_info({}, rollingSplits_df, experiment_name="exp-004")


ROLLING ORIGIN EVALUATION
Total folds: 35
Frequency: MS
Data range: 2014-01-03 to 2016-12-30

Fold Summary:
  Fold 1: Train [2014-01-03 to 2014-02-01] (297 records) → Val [2014-02-02 to 2014-02-28] (186 records)
  Fold 2: Train [2014-01-03 to 2014-03-01] (494 records) → Val [2014-03-02 to 2014-03-31] (319 records)
  Fold 3: Train [2014-01-03 to 2014-04-01] (835 records) → Val [2014-04-02 to 2014-04-30] (308 records)
  Fold 4: Train [2014-01-03 to 2014-05-01] (1,165 records) → Val [2014-05-02 to 2014-05-31] (319 records)
  Fold 5: Train [2014-01-03 to 2014-06-01] (1,506 records) → Val [2014-06-02 to 2014-06-30] (308 records)
  Fold 6: Train [2014-01-03 to 2014-07-01] (1,836 records) → Val [2014-07-02 to 2014-07-31] (319 records)
  Fold 7: Train [2014-01-03 to 2014-08-01] (2,177 records) → Val [2014-08-02 to 2014-08-31] (319 records)
  Fold 8: Train [2014-01-03 to 2014-09-01] (2,518 records) → Val [2014-09-02 to 2014-09-30] (308 records)
  Fold 9: Train [2014-01-03 to 2014-10-01] (2,848

In [9]:
# 2. Rolling Origin Split (More Suitable in TimeSeries) like K-Folds
from src.data_preparation import rolling_origin_evaluation,save_splits_info

rollingSplits_df2 = rolling_origin_evaluation(df=df2, start_train = '2014-01-2',
    end_train = '2016-12-31')
save_splits_info({}, rollingSplits_df2, experiment_name="exp-005")


ROLLING ORIGIN EVALUATION
Total folds: 35
Frequency: MS
Data range: 2014-01-03 to 2016-12-30

Fold Summary:
  Fold 1: Train [2014-01-03 to 2014-02-01] (270 records) → Val [2014-02-02 to 2014-02-28] (170 records)
  Fold 2: Train [2014-01-03 to 2014-03-01] (450 records) → Val [2014-03-02 to 2014-03-31] (290 records)
  Fold 3: Train [2014-01-03 to 2014-04-01] (760 records) → Val [2014-04-02 to 2014-04-30] (280 records)
  Fold 4: Train [2014-01-03 to 2014-05-01] (1,060 records) → Val [2014-05-02 to 2014-05-31] (290 records)
  Fold 5: Train [2014-01-03 to 2014-06-01] (1,370 records) → Val [2014-06-02 to 2014-06-30] (280 records)
  Fold 6: Train [2014-01-03 to 2014-07-01] (1,670 records) → Val [2014-07-02 to 2014-07-31] (290 records)
  Fold 7: Train [2014-01-03 to 2014-08-01] (1,980 records) → Val [2014-08-02 to 2014-08-31] (290 records)
  Fold 8: Train [2014-01-03 to 2014-09-01] (2,290 records) → Val [2014-09-02 to 2014-09-30] (280 records)
  Fold 9: Train [2014-01-03 to 2014-10-01] (2,590

In [10]:
import pandas as pd
import numpy as np
df_noIndex = df.reset_index()
df_noIndex['nam_target_time'] = pd.to_datetime(df_noIndex['nam_target_time'])
df_noIndex['measurement_time'] = pd.to_datetime(df_noIndex['measurement_time'])
df_noIndex['time_diff'] = (df_noIndex['nam_target_time'] - df_noIndex['measurement_time']).dt.total_seconds() / 3600  # Calculate the time difference in hours
    
# Check if the time difference is exactly 1 hour
is_nam_lag_correct = np.allclose(df_noIndex['time_diff'], 1)  # All time differences should be 1 hour
print(f"Does NAM records have a 1-hour lag ahead from measurements? {'Yes' if is_nam_lag_correct else 'No'}")

Does NAM records have a 1-hour lag ahead from measurements? Yes


In [11]:
# # 2. Check if sequences have the same length across each day
# # Extract date from the 'measurement_time'
# df_noIndex['date'] = df_noIndex['measurement_time'].dt.date
    
# # Group by date and count records per day
# daily_counts = df_noIndex.groupby('date').size()
    
# # Find dates where the record count is not equal to 24 (assuming 24 records per day)
# inconsistent_dates = daily_counts[daily_counts != 11] 
    
# # Print dates where records differ from 24
# if not inconsistent_dates.empty:
#     print("Dates with record counts different from 14 hours:")
#     print(inconsistent_dates)
#     print(len(inconsistent_dates))
# else:
#     print("All dates have the expected 14-hour record count.")

In [12]:
# # Create a list of dates to drop
# dates_to_drop = [
#     '2014-01-03', '2014-01-30', '2014-02-05', '2014-02-13', '2014-02-19',
#     '2015-08-23', '2015-08-24', '2015-12-31', '2016-01-01', '2016-07-13',
#     '2016-07-15', '2016-12-31'
# ]

# # Print the first few rows of the index to see its format
# print("Index sample before processing:")
# print(df.index[:5])
# print("\nIndex timezone:", df.index.tz)

# # Convert dates to datetime with UTC timezone
# dates_to_drop = pd.to_datetime(dates_to_drop).tz_localize('UTC')

# # Get the original size of the DataFrame
# original_size = len(df)

# df = df[~df.index.normalize().isin(dates_to_drop)]

# # Print the number of rows dropped
# print(f"\nOriginal number of records: {original_size}")
# print(f"Number of records after dropping dates: {len(df)}")
# print(f"Number of records dropped: {original_size - len(df)}")

In [13]:
df.columns

Index(['ghi', 'dni', 'PST_Time_meas', 'solar_zenith', 'time_gap_hours',
       'time_gap_norm', 'day_boundary_flag', 'hour_progression',
       'absolute_hour', 'GHI_cs', 'DNI_cs', 'CSI_ghi', 'CSI_dni',
       'season_flag', 'hour_sin', 'hour_cos', 'month_sin', 'month_cos',
       'nam_ghi', 'nam_dni', 'nam_cc', 'PST_Time_nam', 'nam_target_time',
       'B_CSI_ghi_8h', 'V_CSI_ghi_8h', 'L_CSI_ghi_8h', 'B_CSI_ghi_9h',
       'V_CSI_ghi_9h', 'L_CSI_ghi_9h', 'B_CSI_ghi_10h', 'V_CSI_ghi_10h',
       'L_CSI_ghi_10h', 'B_CSI_ghi_11h', 'V_CSI_ghi_11h', 'L_CSI_ghi_11h',
       'B_CSI_ghi_12h', 'V_CSI_ghi_12h', 'L_CSI_ghi_12h', 'B_CSI_ghi_13h',
       'V_CSI_ghi_13h', 'L_CSI_ghi_13h', 'B_CSI_ghi_14h', 'V_CSI_ghi_14h',
       'L_CSI_ghi_14h', 'B_CSI_ghi_15h', 'V_CSI_ghi_15h', 'L_CSI_ghi_15h',
       'B_CSI_ghi_16h', 'V_CSI_ghi_16h', 'L_CSI_ghi_16h', 'B_CSI_ghi_17h',
       'V_CSI_ghi_17h', 'L_CSI_ghi_17h', 'B_CSI_ghi_18h', 'V_CSI_ghi_18h',
       'L_CSI_ghi_18h', 'B_CSI_ghi_19h', 'V_CSI_ghi_19h',

In [14]:



features_to_check = ['ghi', 'dni', 'solar_zenith', 'GHI_cs', 'DNI_cs', 'CSI_ghi', 
                                   'CSI_dni',
                                   'nam_ghi', 'nam_dni', 'nam_cc', 'B_CSI_ghi_8h', 'V_CSI_ghi_8h', 'L_CSI_ghi_8h', 'B_CSI_ghi_9h',
       'V_CSI_ghi_9h', 'L_CSI_ghi_9h', 'B_CSI_ghi_10h', 'V_CSI_ghi_10h',
       'L_CSI_ghi_10h', 'B_CSI_ghi_11h', 'V_CSI_ghi_11h', 'L_CSI_ghi_11h',
       'B_CSI_ghi_12h', 'V_CSI_ghi_12h', 'L_CSI_ghi_12h', 'B_CSI_ghi_13h',
       'V_CSI_ghi_13h', 'L_CSI_ghi_13h', 'B_CSI_ghi_14h', 'V_CSI_ghi_14h',
       'L_CSI_ghi_14h', 'B_CSI_ghi_15h', 'V_CSI_ghi_15h', 'L_CSI_ghi_15h',
       'B_CSI_ghi_16h', 'V_CSI_ghi_16h', 'L_CSI_ghi_16h', 'B_CSI_ghi_17h',
       'V_CSI_ghi_17h', 'L_CSI_ghi_17h', 'B_CSI_ghi_18h', 'V_CSI_ghi_18h',
       'L_CSI_ghi_18h', 'B_CSI_ghi_19h', 'V_CSI_ghi_19h', 'L_CSI_ghi_19h',
       '80_dwsw', '80_cloud_cover', '56_dwsw', '56_cloud_cover',
       '20_dwsw', '20_cloud_cover', '88_dwsw', '88_cloud_cover', 'AVG(R)',
       'STD(R)', 'ENT(R)', 'AVG(G)', 'STD(G)', 'ENT(G)', 'AVG(B)', 'STD(B)',
       'ENT(B)', 'AVG(RB)', 'STD(RB)', 'ENT(RB)', 'AVG(NRB)', 'STD(NRB)',
       'ENT(NRB)']

# 2. Calculate the number of NaN values for each feature
# We use .isna() instead of == 0
nans_per_feature = df[features_to_check].isna().sum()

# 3. Filter to get only features that actually have NaN values
features_with_nans = nans_per_feature[nans_per_feature > 0]

# 4. Report the findings for which features have NaNs
if features_with_nans.empty:
    print("No NaN (missing) values found in any of the specified features.")
else:
    print("--- Features With NaN Values ---")
    print("The following features have NaN values, with the total count for each:")
    # Sort for clearer output
    print(features_with_nans.sort_values(ascending=False))
    print("\n" + "="*40 + "\n")

    # 5. Analyze the distribution of hours for rows containing NaNs
    print("--- Distribution of NaN-Value Records by Hour ---")
    
    # Create a boolean mask for rows that contain *at least one* NaN
    # in the specified columns
    rows_with_any_nan = df[features_to_check].isna().any(axis=1)
    
    if rows_with_any_nan.sum() > 0:
        # Get the index for these rows
        nan_rows_index = df.index[rows_with_any_nan]
        
        # Extract the hour from the DatetimeIndex and get the value counts
        hour_distribution = nan_rows_index.hour.value_counts().sort_index()
        
        print("Distribution of records (rows) containing at least one NaN, by hour:")
        print(hour_distribution)
        
        # Optional: Print total number of affected rows
        print(f"\nTotal number of rows with at least one NaN: {rows_with_any_nan.sum()}")
    else:
        # This case shouldn't be hit if features_with_nans was not empty,
        # but it's good practice to include.
        print("No rows found with NaN values (this is unexpected, check logic).")

--- Features With NaN Values ---
The following features have NaN values, with the total count for each:
AVG(R)      1259
STD(R)      1259
ENT(R)      1259
AVG(G)      1259
STD(G)      1259
ENT(G)      1259
AVG(B)      1259
STD(B)      1259
ENT(B)      1259
AVG(RB)     1259
STD(RB)     1259
ENT(RB)     1259
AVG(NRB)    1259
STD(NRB)    1259
ENT(NRB)    1259
dtype: int64


--- Distribution of NaN-Value Records by Hour ---
Distribution of records (rows) containing at least one NaN, by hour:
measurement_time
2     1050
14      84
15      14
16      14
17      14
18      14
19      14
20      13
21      14
22      14
23      14
Name: count, dtype: int64

Total number of rows with at least one NaN: 1259


In [15]:
df_interpolated = df.interpolate(method='linear')

/tmp/ipykernel_2893040/931910069.py:1: FutureWarning: DataFrame.interpolate with object dtype is deprecated and will raise in a future version. Call obj.infer_objects(copy=False) before interpolating instead.
  df_interpolated = df.interpolate(method='linear')


In [16]:
df_interpolated2 = df2.interpolate(method='linear')


/tmp/ipykernel_2893040/1599088726.py:1: FutureWarning: DataFrame.interpolate with object dtype is deprecated and will raise in a future version. Call obj.infer_objects(copy=False) before interpolating instead.
  df_interpolated2 = df2.interpolate(method='linear')


In [17]:



features_to_check = ['ghi', 'dni', 'solar_zenith', 'GHI_cs', 'DNI_cs', 'CSI_ghi', 
                                   'CSI_dni',
                                   'nam_ghi', 'nam_dni', 'nam_cc', 'B_CSI_ghi_8h', 'V_CSI_ghi_8h', 'L_CSI_ghi_8h', 'B_CSI_ghi_9h',
       'V_CSI_ghi_9h', 'L_CSI_ghi_9h', 'B_CSI_ghi_10h', 'V_CSI_ghi_10h',
       'L_CSI_ghi_10h', 'B_CSI_ghi_11h', 'V_CSI_ghi_11h', 'L_CSI_ghi_11h',
       'B_CSI_ghi_12h', 'V_CSI_ghi_12h', 'L_CSI_ghi_12h', 'B_CSI_ghi_13h',
       'V_CSI_ghi_13h', 'L_CSI_ghi_13h', 'B_CSI_ghi_14h', 'V_CSI_ghi_14h',
       'L_CSI_ghi_14h', 'B_CSI_ghi_15h', 'V_CSI_ghi_15h', 'L_CSI_ghi_15h',
       'B_CSI_ghi_16h', 'V_CSI_ghi_16h', 'L_CSI_ghi_16h', 'B_CSI_ghi_17h',
       'V_CSI_ghi_17h', 'L_CSI_ghi_17h', 'B_CSI_ghi_18h', 'V_CSI_ghi_18h',
       'L_CSI_ghi_18h', 'B_CSI_ghi_19h', 'V_CSI_ghi_19h', 'L_CSI_ghi_19h',
       '80_dwsw', '80_cloud_cover', '56_dwsw', '56_cloud_cover',
       '20_dwsw', '20_cloud_cover', '88_dwsw', '88_cloud_cover', 'AVG(R)',
       'STD(R)', 'ENT(R)', 'AVG(G)', 'STD(G)', 'ENT(G)', 'AVG(B)', 'STD(B)',
       'ENT(B)', 'AVG(RB)', 'STD(RB)', 'ENT(RB)', 'AVG(NRB)', 'STD(NRB)',
       'ENT(NRB)']

# 2. Calculate the number of NaN values for each feature
# We use .isna() instead of == 0
nans_per_feature = df_interpolated[features_to_check].isna().sum()

# 3. Filter to get only features that actually have NaN values
features_with_nans = nans_per_feature[nans_per_feature > 0]

# 4. Report the findings for which features have NaNs
if features_with_nans.empty:
    print("No NaN (missing) values found in any of the specified features.")
else:
    print("--- Features With NaN Values ---")
    print("The following features have NaN values, with the total count for each:")
    # Sort for clearer output
    print(features_with_nans.sort_values(ascending=False))
    print("\n" + "="*40 + "\n")

    # 5. Analyze the distribution of hours for rows containing NaNs
    print("--- Distribution of NaN-Value Records by Hour ---")
    
    # Create a boolean mask for rows that contain *at least one* NaN
    # in the specified columns
    rows_with_any_nan = df_interpolated[features_to_check].isna().any(axis=1)
    
    if rows_with_any_nan.sum() > 0:
        # Get the index for these rows
        nan_rows_index = df_interpolated.index[rows_with_any_nan]
        
        # Extract the hour from the DatetimeIndex and get the value counts
        hour_distribution = nan_rows_index.hour.value_counts().sort_index()
        
        print("Distribution of records (rows) containing at least one NaN, by hour:")
        print(hour_distribution)
        
        # Optional: Print total number of affected rows
        print(f"\nTotal number of rows with at least one NaN: {rows_with_any_nan.sum()}")
    else:
        # This case shouldn't be hit if features_with_nans was not empty,
        # but it's good practice to include.
        print("No rows found with NaN values (this is unexpected, check logic).")

--- Features With NaN Values ---
The following features have NaN values, with the total count for each:
AVG(R)      1
STD(R)      1
ENT(R)      1
AVG(G)      1
STD(G)      1
ENT(G)      1
AVG(B)      1
STD(B)      1
ENT(B)      1
AVG(RB)     1
STD(RB)     1
ENT(RB)     1
AVG(NRB)    1
STD(NRB)    1
ENT(NRB)    1
dtype: int64


--- Distribution of NaN-Value Records by Hour ---
Distribution of records (rows) containing at least one NaN, by hour:
measurement_time
14    1
Name: count, dtype: int64

Total number of rows with at least one NaN: 1


----

### Data Preparation for Experiments

In [18]:
import torch
device = (
    "cuda"
    if torch.cuda.is_available()
    else "mps"
    if hasattr(torch.backends, "mps") and torch.backends.mps.is_available()
    else "cpu"
)

device

'cuda'

In [19]:
df_interpolated2.columns

Index(['ghi', 'dni', 'PST_Time_meas', 'solar_zenith', 'time_gap_hours',
       'time_gap_norm', 'day_boundary_flag', 'hour_progression',
       'absolute_hour', 'GHI_cs', 'DNI_cs', 'CSI_ghi', 'CSI_dni',
       'season_flag', 'hour_sin', 'hour_cos', 'month_sin', 'month_cos',
       'nam_ghi', 'nam_dni', 'nam_cc', 'PST_Time_nam', 'nam_target_time',
       'B_CSI_ghi_8h', 'V_CSI_ghi_8h', 'L_CSI_ghi_8h', 'B_CSI_ghi_9h',
       'V_CSI_ghi_9h', 'L_CSI_ghi_9h', 'B_CSI_ghi_10h', 'V_CSI_ghi_10h',
       'L_CSI_ghi_10h', 'B_CSI_ghi_11h', 'V_CSI_ghi_11h', 'L_CSI_ghi_11h',
       'B_CSI_ghi_12h', 'V_CSI_ghi_12h', 'L_CSI_ghi_12h', 'B_CSI_ghi_13h',
       'V_CSI_ghi_13h', 'L_CSI_ghi_13h', 'B_CSI_ghi_14h', 'V_CSI_ghi_14h',
       'L_CSI_ghi_14h', 'B_CSI_ghi_15h', 'V_CSI_ghi_15h', 'L_CSI_ghi_15h',
       'B_CSI_ghi_16h', 'V_CSI_ghi_16h', 'L_CSI_ghi_16h', 'B_CSI_ghi_17h',
       'V_CSI_ghi_17h', 'L_CSI_ghi_17h', 'B_CSI_ghi_18h', 'V_CSI_ghi_18h',
       'L_CSI_ghi_18h', 'B_CSI_ghi_19h', 'V_CSI_ghi_19h',

In [20]:
df_interpolated.columns

Index(['ghi', 'dni', 'PST_Time_meas', 'solar_zenith', 'time_gap_hours',
       'time_gap_norm', 'day_boundary_flag', 'hour_progression',
       'absolute_hour', 'GHI_cs', 'DNI_cs', 'CSI_ghi', 'CSI_dni',
       'season_flag', 'hour_sin', 'hour_cos', 'month_sin', 'month_cos',
       'nam_ghi', 'nam_dni', 'nam_cc', 'PST_Time_nam', 'nam_target_time',
       'B_CSI_ghi_8h', 'V_CSI_ghi_8h', 'L_CSI_ghi_8h', 'B_CSI_ghi_9h',
       'V_CSI_ghi_9h', 'L_CSI_ghi_9h', 'B_CSI_ghi_10h', 'V_CSI_ghi_10h',
       'L_CSI_ghi_10h', 'B_CSI_ghi_11h', 'V_CSI_ghi_11h', 'L_CSI_ghi_11h',
       'B_CSI_ghi_12h', 'V_CSI_ghi_12h', 'L_CSI_ghi_12h', 'B_CSI_ghi_13h',
       'V_CSI_ghi_13h', 'L_CSI_ghi_13h', 'B_CSI_ghi_14h', 'V_CSI_ghi_14h',
       'L_CSI_ghi_14h', 'B_CSI_ghi_15h', 'V_CSI_ghi_15h', 'L_CSI_ghi_15h',
       'B_CSI_ghi_16h', 'V_CSI_ghi_16h', 'L_CSI_ghi_16h', 'B_CSI_ghi_17h',
       'V_CSI_ghi_17h', 'L_CSI_ghi_17h', 'B_CSI_ghi_18h', 'V_CSI_ghi_18h',
       'L_CSI_ghi_18h', 'B_CSI_ghi_19h', 'V_CSI_ghi_19h',

In [23]:
df_phase1 = df_interpolated[['solar_zenith', 'CSI_ghi','time_gap_hours',
       'time_gap_norm', 'day_boundary_flag', 'hour_progression',
       'absolute_hour', 'PST_Time_meas',
       'season_flag', 'hour_sin', 'hour_cos', 'month_sin', 'month_cos']]


df_phase2 = df_interpolated[['solar_zenith', 'CSI_ghi','time_gap_hours',
       'time_gap_norm', 'day_boundary_flag', 'hour_progression',
       'absolute_hour',
       'season_flag', 'hour_sin', 'hour_cos', 'month_sin', 'month_cos', 'B_CSI_ghi_8h', 'V_CSI_ghi_8h', 'L_CSI_ghi_8h', 'B_CSI_ghi_9h',
       'V_CSI_ghi_9h', 'L_CSI_ghi_9h', 'B_CSI_ghi_10h', 'V_CSI_ghi_10h',
       'L_CSI_ghi_10h', 'B_CSI_ghi_11h', 'V_CSI_ghi_11h', 'L_CSI_ghi_11h',
       'B_CSI_ghi_12h', 'V_CSI_ghi_12h', 'L_CSI_ghi_12h', 'B_CSI_ghi_13h',
       'V_CSI_ghi_13h', 'L_CSI_ghi_13h', 'B_CSI_ghi_14h', 'V_CSI_ghi_14h',
       'L_CSI_ghi_14h', 'B_CSI_ghi_15h', 'V_CSI_ghi_15h', 'L_CSI_ghi_15h',
       'B_CSI_ghi_16h', 'V_CSI_ghi_16h', 'L_CSI_ghi_16h', 'B_CSI_ghi_17h',
       'V_CSI_ghi_17h', 'L_CSI_ghi_17h', 'B_CSI_ghi_18h', 'V_CSI_ghi_18h',
       'L_CSI_ghi_18h', 'B_CSI_ghi_19h', 'V_CSI_ghi_19h', 'L_CSI_ghi_19h']]

df_phase3 = df_interpolated[['solar_zenith', 'CSI_ghi','time_gap_hours',
       'time_gap_norm', 'day_boundary_flag', 'hour_progression',
       'absolute_hour',
       'season_flag', 'hour_sin', 'hour_cos', 'month_sin', 'month_cos', 'B_CSI_ghi_8h', 'V_CSI_ghi_8h', 'L_CSI_ghi_8h', 'B_CSI_ghi_9h',
       'V_CSI_ghi_9h', 'L_CSI_ghi_9h', 'B_CSI_ghi_10h', 'V_CSI_ghi_10h',
       'L_CSI_ghi_10h', 'B_CSI_ghi_11h', 'V_CSI_ghi_11h', 'L_CSI_ghi_11h',
       'B_CSI_ghi_12h', 'V_CSI_ghi_12h', 'L_CSI_ghi_12h', 'B_CSI_ghi_13h',
       'V_CSI_ghi_13h', 'L_CSI_ghi_13h', 'B_CSI_ghi_14h', 'V_CSI_ghi_14h',
       'L_CSI_ghi_14h', 'B_CSI_ghi_15h', 'V_CSI_ghi_15h', 'L_CSI_ghi_15h',
       'B_CSI_ghi_16h', 'V_CSI_ghi_16h', 'L_CSI_ghi_16h', 'B_CSI_ghi_17h',
       'V_CSI_ghi_17h', 'L_CSI_ghi_17h', 'B_CSI_ghi_18h', 'V_CSI_ghi_18h',
       'L_CSI_ghi_18h', 'B_CSI_ghi_19h', 'V_CSI_ghi_19h', 'L_CSI_ghi_19h',
       'nam_ghi', 'nam_cc']]

df_phase4 = df_interpolated[['solar_zenith', 'CSI_ghi','time_gap_hours',
       'time_gap_norm', 'day_boundary_flag', 'hour_progression',
       'absolute_hour',
       'season_flag', 'hour_sin', 'hour_cos', 'month_sin', 'month_cos', 'B_CSI_ghi_8h', 'V_CSI_ghi_8h', 'L_CSI_ghi_8h', 'B_CSI_ghi_9h',
       'V_CSI_ghi_9h', 'L_CSI_ghi_9h', 'B_CSI_ghi_10h', 'V_CSI_ghi_10h',
       'L_CSI_ghi_10h', 'B_CSI_ghi_11h', 'V_CSI_ghi_11h', 'L_CSI_ghi_11h',
       'B_CSI_ghi_12h', 'V_CSI_ghi_12h', 'L_CSI_ghi_12h', 'B_CSI_ghi_13h',
       'V_CSI_ghi_13h', 'L_CSI_ghi_13h', 'B_CSI_ghi_14h', 'V_CSI_ghi_14h',
       'L_CSI_ghi_14h', 'B_CSI_ghi_15h', 'V_CSI_ghi_15h', 'L_CSI_ghi_15h',
       'B_CSI_ghi_16h', 'V_CSI_ghi_16h', 'L_CSI_ghi_16h', 'B_CSI_ghi_17h',
       'V_CSI_ghi_17h', 'L_CSI_ghi_17h', 'B_CSI_ghi_18h', 'V_CSI_ghi_18h',
       'L_CSI_ghi_18h', 'B_CSI_ghi_19h', 'V_CSI_ghi_19h', 'L_CSI_ghi_19h',
       'nam_ghi', 'nam_cc', '80_dwsw', 'valtime', '80_cloud_cover', '56_dwsw', '56_cloud_cover',
       '20_dwsw', '20_cloud_cover', '88_dwsw', '88_cloud_cover']]

df_phase5 = df_interpolated[['solar_zenith', 'CSI_ghi','time_gap_hours',
       'time_gap_norm', 'day_boundary_flag', 'hour_progression',
       'absolute_hour',
       'season_flag', 'hour_sin', 'hour_cos', 'month_sin', 'month_cos', 'B_CSI_ghi_8h', 'V_CSI_ghi_8h', 'L_CSI_ghi_8h', 'B_CSI_ghi_9h',
       'V_CSI_ghi_9h', 'L_CSI_ghi_9h', 'B_CSI_ghi_10h', 'V_CSI_ghi_10h',
       'L_CSI_ghi_10h', 'B_CSI_ghi_11h', 'V_CSI_ghi_11h', 'L_CSI_ghi_11h',
       'B_CSI_ghi_12h', 'V_CSI_ghi_12h', 'L_CSI_ghi_12h', 'B_CSI_ghi_13h',
       'V_CSI_ghi_13h', 'L_CSI_ghi_13h', 'B_CSI_ghi_14h', 'V_CSI_ghi_14h',
       'L_CSI_ghi_14h', 'B_CSI_ghi_15h', 'V_CSI_ghi_15h', 'L_CSI_ghi_15h',
       'B_CSI_ghi_16h', 'V_CSI_ghi_16h', 'L_CSI_ghi_16h', 'B_CSI_ghi_17h',
       'V_CSI_ghi_17h', 'L_CSI_ghi_17h', 'B_CSI_ghi_18h', 'V_CSI_ghi_18h',
       'L_CSI_ghi_18h', 'B_CSI_ghi_19h', 'V_CSI_ghi_19h', 'L_CSI_ghi_19h',
       'nam_ghi', 'nam_cc','80_cloud_cover',  '56_cloud_cover',
        '20_cloud_cover', '88_cloud_cover']]

df_phase6 = df_interpolated[['solar_zenith', 'CSI_ghi','time_gap_hours',
       'time_gap_norm', 'day_boundary_flag', 'hour_progression',
       'absolute_hour',
       'season_flag', 'hour_sin', 'hour_cos', 'month_sin', 'month_cos', 'B_CSI_ghi_8h', 'V_CSI_ghi_8h', 'L_CSI_ghi_8h', 'B_CSI_ghi_9h',
       'V_CSI_ghi_9h', 'L_CSI_ghi_9h', 'B_CSI_ghi_10h', 'V_CSI_ghi_10h',
       'L_CSI_ghi_10h', 'B_CSI_ghi_11h', 'V_CSI_ghi_11h', 'L_CSI_ghi_11h',
       'B_CSI_ghi_12h', 'V_CSI_ghi_12h', 'L_CSI_ghi_12h', 'B_CSI_ghi_13h',
       'V_CSI_ghi_13h', 'L_CSI_ghi_13h', 'B_CSI_ghi_14h', 'V_CSI_ghi_14h',
       'L_CSI_ghi_14h', 'B_CSI_ghi_15h', 'V_CSI_ghi_15h', 'L_CSI_ghi_15h',
       'B_CSI_ghi_16h', 'V_CSI_ghi_16h', 'L_CSI_ghi_16h', 'B_CSI_ghi_17h',
       'V_CSI_ghi_17h', 'L_CSI_ghi_17h', 'B_CSI_ghi_18h', 'V_CSI_ghi_18h',
       'L_CSI_ghi_18h', 'B_CSI_ghi_19h', 'V_CSI_ghi_19h', 'L_CSI_ghi_19h',
       'nam_ghi', 'nam_cc', '80_dwsw', 'valtime', '80_cloud_cover', '56_dwsw', '56_cloud_cover',
       '20_dwsw', '20_cloud_cover', '88_dwsw', '88_cloud_cover'
      ]]


-----

## Testing and Validating the Data Preparation for traning

In [24]:
from src.preprocessing import build_model_arrays, to_fixedgrid_multiindex

TARGET_COL = "CSI_ghi"
timestamp_col="measurement_time"

df_phase1 = df_interpolated[['solar_zenith', 'CSI_ghi',
       'absolute_hour',
       'season_flag' ]]


feature_cols = [c for c in df_phase1.columns.tolist() if c != TARGET_COL]
# fixed_df = to_fixedgrid_multiindex(df_phase1, timestamp_col="measurement_time", expected_T=None) 
test_df = df_phase1.copy(deep=True)

test_df.reset_index(inplace=True)
test_df["UTC_date"] = test_df[timestamp_col].dt.tz_convert("UTC").dt.date
test_df["PST_date"] = (
    test_df[timestamp_col]
    .dt.tz_convert("US/Pacific")  # Converts UTC to Pacific Time (handles DST automatically)
    .dt.date
)
test_df["bin_id"] = test_df.groupby("PST_date").cumcount()
test_df = test_df.set_index(["PST_date", "bin_id"]).sort_index()


# ph1_X, ph1_Y, ph1_labels_list = build_model_arrays(
#         test_df,
#         feature_cols=feature_cols,  # <-- This is correct!
#         target_col=TARGET_COL,
#         history_days=2,
#         horizon_days=1,
#     )




In [25]:
test_df

measurement_time  solar_zenith   CSI_ghi  \
PST_date   bin_id                                                     
2014-01-03 0      2014-01-03 14:00:00+00:00    100.276046  0.000000   
           1      2014-01-03 15:00:00+00:00     89.724162  1.200000   
           2      2014-01-03 16:00:00+00:00     80.146724  1.200000   
           3      2014-01-03 17:00:00+00:00     71.988814  1.200000   
           4      2014-01-03 18:00:00+00:00     65.816158  1.200000   
...                                     ...           ...       ...   
2016-12-30 5      2016-12-30 19:00:00+00:00     62.502377  1.051822   
           6      2016-12-30 20:00:00+00:00     62.079248  1.001250   
           7      2016-12-30 21:00:00+00:00     64.765899  0.961864   
           8      2016-12-30 22:00:00+00:00     70.203098  0.896756   
           9      2016-12-30 23:00:00+00:00     77.805433  0.804137   

                   absolute_hour  season_flag    UTC_date  
PST_date   bin_id                                          
2014-01-03 0           24.491667            2  2014-01-03  
           1           25.491667            2  2014-01-03  
           2           26.491667            2  2014-01-03  
           3           27.491667            2  2014-01-03  
           4           28.491667            2  2014-01-03  
...                          ...          ...         ...  
2016-12-30 5        26237.491667            2  2016-12-30  
           6        26238.491667            2  2016-12-30  
           7        26239.491667            2  2016-12-30  
           8        26240.491667            2  2016-12-30  
           9        26241.491667            2  2016-12-30  

[11560 rows x 6 columns]

In [26]:
dates = list(test_df.index.get_level_values("PST_date").unique())
K = int(test_df.index.get_level_values("bin_id").max()) + 1
F = len(feature_cols)
X = np.full((len(dates), K, F), np.nan, dtype=float)

for j, col in enumerate(feature_cols):
    if col not in test_df.columns:
        continue

    

    x_val = test_df['solar_zenith'].groupby(level=["PST_date", "bin_id"]).first().unstack("bin_id").reindex(index=dates, columns=range(K))
    X[:, :, j] = x_val.values 
    

In [27]:
len(dates), X.shape

(1051, (1051, 11, 3))

In [28]:
# Target Tensor Building
f_cols = ['CSI_ghi']

dates = list(test_df.index.get_level_values("PST_date").unique())
K = int(test_df.index.get_level_values("bin_id").max()) + 1
F = len(f_cols)
Y = np.full((len(dates), K, F), np.nan, dtype=float)

for j, col in enumerate(f_cols):
    if col not in test_df.columns:
        continue

    

    y_val = test_df['solar_zenith'].groupby(level=["PST_date", "bin_id"]).first().unstack("bin_id").reindex(index=dates, columns=range(K))
    Y[:, :, j] = x_val.values 
    

In [29]:
Y = Y[:,:,0]

In [30]:
len(dates), Y.shape

(1051, (1051, 11))

In [31]:
history_days = 3
horizon_days = 1
num_days = X.shape[0]   # Avaible dates
K = X.shape[1]          # Hours records
F = X.shape[2]          # Avaible Features

samples = max(0, num_days - history_days - horizon_days + 1)
X_list, Y_list, labels = [], [], []
for i in range(samples):
    past = slice(i, i+history_days)
    fut  = slice(i+history_days, i+history_days+horizon_days)
    X_i = X[past,:,:]
    Y_i = Y[fut,:]
    X_list.append(X_i)
    Y_list.append(Y_i)
    labels.append(dates[i+history_days+horizon_days-1])


 # converts the list of individual samples into the final, single, massive tensor required by the deep learning model.
X_tensor = np.stack(X_list, axis=0) if X_list else np.empty((0,history_days,K,F))
Y_tensor = np.stack(Y_list, axis=0) if Y_list else np.empty((0,horizon_days,K))
# X: A 4D tensor of shape (samples, history_days, K, F)



In [34]:
X_tensor.shape, Y_tensor.shape, len(labels), labels[30:36]

((1048, 3, 11, 3),
 (1048, 1, 11),
 1048,
 [datetime.date(2014, 2, 11),
  datetime.date(2014, 2, 12),
  datetime.date(2014, 2, 19),
  datetime.date(2014, 2, 20),
  datetime.date(2014, 2, 21),
  datetime.date(2014, 2, 22)])

In [42]:
labels.index

<function list.index(value, start=0, stop=9223372036854775807, /)>

In [35]:
from src.utils import DataManager

data_manager = DataManager()

In [ ]:

data_manager.save_arrays(
    X_tensor, Y_tensor,
    pd.DataFrame(
        index=pd.to_datetime(labels, utc=True)),
    filename_prefix='Test_dataFile_prefix',  # <-- You are saving as 'phas2_data'
    feature_cols=feature_cols,
    target_col=TARGET_COL,
    metadata={
        "input_csv":       "data/processed/df_1h_lag_BLV_spatial_images.csv",
        "timestamp_col": "measurement_time",
        "feature_set": feature_cols,
        "history_days": 7,
        "horizon_days": 1,
        "k_bins": None,
    }
)



INFO:src.utils:Saved arrays to data/Test_dataFile_prefix_*.npy
INFO:src.utils:X shape: (1048, 3, 11, 3), Y shape: (1048, 1, 11)


In [47]:
feature_cols, TARGET_COL, labels.index

(['solar_zenith', 'absolute_hour', 'season_flag'],
 'CSI_ghi',
 <function list.index(value, start=0, stop=9223372036854775807, /)>)

In [55]:
# --- Block 4: Configure and Run ---
LSTM_CONFIG = {
    "experiment_name": "TESTEXP_UniLSTM_p1_exp01_essentialF",
    "model_type": "LSTM",
    "model_config": {
        "hidden_size": 64,
        "num_layers": 2,
        "dropout": 0.35,
        "bidirectional": True,
    },
    "data_prefix": "phas1_data",  # <-- FIX 2: Match the filename_prefix
    "splits_file": "exp-004/exp-004rolling_origin_splits.json",
    "feature_cols": feature_cols,
    "feature_selection": feature_cols, # <-- This is correct!
    "target_col": TARGET_COL,
    "batch_size": 32,
    "num_epochs": 50,
    "learning_rate": 0.001,
    "loss_function": "Huber",  # <-- This setting needs the fix below
    "early_stopping_patience": 20,
    "max_folds": 3,



}
metadata={
        "input_csv": "data/processed/df_1h_lag_BLV_spatial_images.csv",
        "timestamp_col": "measurement_time",
        "feature_set": feature_cols,
        "history_days": 7,
        "horizon_days": 1,
        "k_bins": None,
    }

expected_features = LSTM_CONFIG.get("feature_cols")

from src.evaluation_utils import build_reference_fixedgrid_frame, SiteLocation

ts_col = metadata.get("timestamp_col", "measurement_time")
csv_path = metadata.get("input_csv")

site_config = SiteLocation(
                latitude=LSTM_CONFIG.get('site_latitude'),
                longitude=LSTM_CONFIG.get('site_longitude'),
                altitude=LSTM_CONFIG.get('site_altitude'),
                timezone=LSTM_CONFIG.get('site_timezone'),
            )
ts_col, csv_path, site_config

reference_df = build_reference_fixedgrid_frame(
                        csv_path,
                        site_config,
                        timestamp_col=ts_col
                    )
reference_df

KeyError: 'target_time'

In [ ]:
from src.pipeline import SolarForecastingPipeline

pipeline = SolarForecastingPipeline(LSTM_CONFIG)
_, summary = pipeline.run()

### End of Data Preparation Validation 
---

In [23]:
from src.preprocessing import KBinConfig, process_splits_to_kbins, build_model_arrays, to_fixedgrid_multiindex
import pandas as pd
from src.utils import DataManager
from src.pipeline import SolarForecastingPipeline

# --- Block 1: Define variables ---
USE_KBINS = False
k_bins = 60
TARGET_COL = "CSI_ghi"  # <-- FIX 1: Define Target FIRST
history_days = 7
horizon_days = 1

# This list now correctly excludes the target
feature_cols = [c for c in df_phase1.columns.tolist() if c != TARGET_COL]

# --- Block 2: Build model arrays (Your code was correct here) ---
if USE_KBINS:
        print("\n--- Step 2: Converting full dataset to K-Bins format ---")
        # ... (rest of your K-Bins code) ...
else:
    print("\n--- Step 2: Building model arrays (X, Y) ---")
    fixed_df = to_fixedgrid_multiindex(df_phase1, timestamp_col="measurement_time", expected_T=None)  # or set T
    
    # You correctly passed the filtered 'feature_cols' here, fixing the data leak!
    ph1_X, ph1_Y, ph1_labels_list = build_model_arrays(
        fixed_df,
        feature_cols=feature_cols,  # <-- This is correct!
        target_col=TARGET_COL,
        history_days=history_days,
        horizon_days=horizon_days,
    )

# --- Block 3: Save arrays ---
data_manager = DataManager()
data_manager.save_arrays(
    ph1_X, ph1_Y,
    pd.DataFrame(index=pd.to_datetime(ph1_labels_list, utc=True)),
    filename_prefix='phas1_data',  # <-- You are saving as 'phas2_data'
    feature_cols=feature_cols,
    target_col=TARGET_COL,
    metadata={
        "input_csv": "data/processed/df_1h_lag_BLV_spatial_images.csv",
        "timestamp_col": "measurement_time",
        "feature_set": feature_cols,
        "history_days": 7,
        "horizon_days": 1,
        "k_bins": None,
    }
)

# --- Block 4: Configure and Run ---
LSTM_CONFIG = {
    "experiment_name": "UniLSTM_p1_exp01_essentialF",
    "model_type": "LSTM",
    "model_config": {
        "hidden_size": 64,
        "num_layers": 2,
        "dropout": 0.35,
        "bidirectional": True,
    },
    "data_prefix": "phas1_data",  # <-- FIX 2: Match the filename_prefix
    "splits_file": "exp-004/exp-004rolling_origin_splits.json",
    "feature_cols": feature_cols,
    "feature_selection": feature_cols, # <-- This is correct!
    "target_col": TARGET_COL,
    "batch_size": 32,
    "num_epochs": 50,
    "learning_rate": 0.001,
    "loss_function": "Huber",  # <-- This setting needs the fix below
    "early_stopping_patience": 20,
    "max_folds": 35,
}

# This will now work after you apply the fixes below
pipeline = SolarForecastingPipeline(LSTM_CONFIG)
_, summary = pipeline.run()

INFO:src.utils:Saved arrays to data/phas1_data_*.npy
INFO:src.utils:X shape: (1055, 7, 11, 11), Y shape: (1055, 1, 11)
INFO:src.pipeline:Loading data...
INFO:src.utils:Loaded arrays from data/phas1_data_*.npy
INFO:src.utils:X shape: (1055, 7, 11, 11), Y shape: (1055, 1, 11)
INFO:src.pipeline:Building FIXED-GRID reference frame (K_bins is None)...



--- Step 2: Building model arrays (X, Y) ---


INFO:src.utils:Loaded 35 folds from exp-004/exp-004rolling_origin_splits.json
INFO:src.pipeline:
=== Running Fold 1 ===
INFO:src.pipeline:Train samples: 21, Val samples: 18
INFO:src.pipeline:Model parameters: 248,715
INFO:src.engine:Using HuberLoss (Smooth L1 Loss)
INFO:src.engine:Epoch [10/50] - Train Loss: 0.076680, Val Loss: 0.104906, Train MAE: 0.300700, Val MAE: 0.391978, LR: 0.001000
INFO:src.engine:Epoch [20/50] - Train Loss: 0.063023, Val Loss: 0.103829, Train MAE: 0.274112, Val MAE: 0.386069, LR: 0.001000
INFO:src.engine:Epoch [30/50] - Train Loss: 0.049382, Val Loss: 0.104260, Train MAE: 0.223215, Val MAE: 0.386043, LR: 0.001000
INFO:src.engine:Epoch [40/50] - Train Loss: 0.043392, Val Loss: 0.097369, Train MAE: 0.207388, Val MAE: 0.379170, LR: 0.001000
INFO:src.engine:Epoch [50/50] - Train Loss: 0.038784, Val Loss: 0.095513, Train MAE: 0.200005, Val MAE: 0.363086, LR: 0.001000
INFO:src.pipeline:Fold 1 - Validation Metrics:
INFO:src.pipeline:  mse: 0.191598
INFO:src.pipeline:

The model has learned the easiest possible trick: it's predicting that the value at the next time step will be the same as the value at the current time step.

It is a Naive (Persistence Model)


"Lazy learning" (the effect) is a model's symptom. "Incorrect sequencing" (the cause) is the disease. Investigation needed.
    - Solution1: Adding time_gap_hours',
       'time_gap_norm', 'day_boundary_flag', 'hour_progression',
       'absolute_hour', to the model, as well the cyclical time features problem is still. 

    - Solution 2 adding more statistical features about historical Lag like 'B_CSI_ghi_8h', 'V_CSI_ghi_8h', 'L_CSI_ghi_8h', 'B_CSI_ghi_9h',
       'V_CSI_ghi_9h', 'L_CSI_ghi_9h', 'B_CSI_ghi_10h', 'V_CSI_ghi_10h',
       'L_CSI_ghi_10h', 'B_CSI_ghi_11h', 'V_CSI_ghi_11h', 'L_CSI_ghi_11h',
       'B_CSI_ghi_12h', 'V_CSI_ghi_12h', 'L_CSI_ghi_12h', 'B_CSI_ghi_13h',
       'V_CSI_ghi_13h', 'L_CSI_ghi_13h', 'B_CSI_ghi_14h', 'V_CSI_ghi_14h',
       'L_CSI_ghi_14h', 'B_CSI_ghi_15h', 'V_CSI_ghi_15h', 'L_CSI_ghi_15h',
       'B_CSI_ghi_16h', 'V_CSI_ghi_16h', 'L_CSI_ghi_16h', 'B_CSI_ghi_17h',
       'V_CSI_ghi_17h', 'L_CSI_ghi_17h', 'B_CSI_ghi_18h', 'V_CSI_ghi_18h',
       'L_CSI_ghi_18h', 'B_CSI_ghi_19h', 'V_CSI_ghi_19h', 'L_CSI_ghi_19h'
    - Problem is still so i will move to the other 



---

In [28]:
from src.preprocessing import KBinConfig, process_splits_to_kbins, build_model_arrays, to_fixedgrid_multiindex
import pandas as pd
from src.utils import DataManager
from src.pipeline import SolarForecastingPipeline

# --- Block 1: Define variables ---
USE_KBINS = False
k_bins = 60
TARGET_COL = "CSI_ghi"  
history_days = 7
horizon_days = 1

# This list now correctly excludes the target
feature_cols = [c for c in df_phase2.columns.tolist() if c != TARGET_COL]

# --- Block 2: Build model arrays (Your code was correct here) ---
if USE_KBINS:
        print("\n--- Step 2: Converting full dataset to K-Bins format ---")
        # ... (rest of your K-Bins code) ...
else:
    print("\n--- Step 2: Building model arrays (X, Y) ---")
    fixed_df = to_fixedgrid_multiindex(df_phase2, timestamp_col="measurement_time", expected_T=None)  # or set T
    
    # You correctly passed the filtered 'feature_cols' here, fixing the data leak!
    ph2_X, ph2_Y, ph2_labels_list = build_model_arrays(
        fixed_df,
        feature_cols=feature_cols,  # <-- This is correct!
        target_col=TARGET_COL,
        history_days=history_days,
        horizon_days=horizon_days,
    )

# --- Block 3: Save arrays ---
data_manager = DataManager()
data_manager.save_arrays(
    ph2_X, ph2_Y,
    pd.DataFrame(index=pd.to_datetime(ph2_labels_list, utc=True)),
    filename_prefix='phas2_data',  # <-- You are saving as 'phas2_data'
    feature_cols=feature_cols,
    target_col=TARGET_COL,
    metadata={
        "input_csv": "data/processed/df_1h_lag_BLV_spatial_images.csv",
        "timestamp_col": "measurement_time",
        "feature_set": feature_cols,
        "history_days": 7,
        "horizon_days": 1,
        "k_bins": None,
    }
)

# --- Block 4: Configure and Run ---
LSTM_CONFIG = {
    "experiment_name": "UniLSTM_p2_exp01_BLVF",
    "model_type": "LSTM",
    "model_config": {
        "hidden_size": 64,
        "num_layers": 2,
        "dropout": 0.35,
        "bidirectional": True,
    },
    "data_prefix": "phas2_data",  # <-- FIX 2: Match the filename_prefix
    "splits_file": "exp-004/exp-004rolling_origin_splits.json",
    "feature_cols": feature_cols,
    "feature_selection": feature_cols, # <-- This is correct!
    "target_col": TARGET_COL,
    "batch_size": 32,
    "num_epochs": 50,
    "learning_rate": 0.001,
    "loss_function": "Huber",  # <-- This setting needs the fix below
    "early_stopping_patience": 20,
    "max_folds": 35,
}

# This will now work after you apply the fixes below
pipeline = SolarForecastingPipeline(LSTM_CONFIG)
_, summary = pipeline.run()

INFO:src.utils:Saved arrays to data/phas2_data_*.npy
INFO:src.utils:X shape: (1055, 7, 11, 47), Y shape: (1055, 1, 11)
INFO:src.pipeline:Loading data...
INFO:src.utils:Loaded arrays from data/phas2_data_*.npy
INFO:src.utils:X shape: (1055, 7, 11, 47), Y shape: (1055, 1, 11)
INFO:src.pipeline:Building FIXED-GRID reference frame (K_bins is None)...



--- Step 2: Building model arrays (X, Y) ---


INFO:src.utils:Loaded 35 folds from exp-004/exp-004rolling_origin_splits.json
INFO:src.pipeline:
=== Running Fold 1 ===
INFO:src.pipeline:Train samples: 21, Val samples: 18
INFO:src.pipeline:Model parameters: 274,059
INFO:src.engine:Using HuberLoss (Smooth L1 Loss)
INFO:src.engine:Epoch [10/50] - Train Loss: 0.069708, Val Loss: 0.099907, Train MAE: 0.282255, Val MAE: 0.374032, LR: 0.001000
INFO:src.engine:Epoch [20/50] - Train Loss: 0.054642, Val Loss: 0.101493, Train MAE: 0.239465, Val MAE: 0.379981, LR: 0.001000
INFO:src.engine:Epoch [30/50] - Train Loss: 0.054945, Val Loss: 0.100272, Train MAE: 0.248461, Val MAE: 0.386173, LR: 0.000500
INFO:src.engine:Early stopping at epoch 31
INFO:src.pipeline:Fold 1 - Validation Metrics:
INFO:src.pipeline:  mse: 0.200547
INFO:src.pipeline:  rmse: 0.447825
INFO:src.pipeline:  mae: 0.386084
INFO:src.pipeline:  r2: 0.083255
INFO:src.pipeline:  mape: 111.452530
INFO:src.pipeline:
=== Running Fold 2 ===
INFO:src.pipeline:Train samples: 40, Val samples

Increasing the Hisory days 

In [ ]:
from src.preprocessing import KBinConfig, process_splits_to_kbins, build_model_arrays, to_fixedgrid_multiindex
import pandas as pd
from src.utils import DataManager
from src.pipeline import SolarForecastingPipeline

# --- Block 1: Define variables ---
USE_KBINS = False
k_bins = 60
TARGET_COL = "CSI_ghi"  
history_days = 7
horizon_days = 1

# This list now correctly excludes the target
feature_cols = [c for c in df_phase2.columns.tolist() if c != TARGET_COL]

# --- Block 2: Build model arrays (Your code was correct here) ---
if USE_KBINS:
        print("\n--- Step 2: Converting full dataset to K-Bins format ---")
        # ... (rest of your K-Bins code) ...
else:
    print("\n--- Step 2: Building model arrays (X, Y) ---")
    fixed_df = to_fixedgrid_multiindex(df_phase2, timestamp_col="measurement_time", expected_T=None)  # or set T
    
    # You correctly passed the filtered 'feature_cols' here, fixing the data leak!
    ph2_X, ph2_Y, ph2_labels_list = build_model_arrays(
        fixed_df,
        feature_cols=feature_cols,  # <-- This is correct!
        target_col=TARGET_COL,
        history_days=history_days,
        horizon_days=horizon_days,
    )

# --- Block 3: Save arrays ---
data_manager = DataManager()
data_manager.save_arrays(
    ph2_X, ph2_Y,
    pd.DataFrame(index=pd.to_datetime(ph2_labels_list, utc=True)),
    filename_prefix='phas2_data',  # <-- You are saving as 'phas2_data'
    feature_cols=feature_cols,
    target_col=TARGET_COL,
    metadata={
        "input_csv": "data/processed/df_1h_lag_BLV_spatial_images.csv",
        "timestamp_col": "measurement_time",
        "feature_set": feature_cols,
        "history_days": 7,
        "horizon_days": 1,
        "k_bins": None,
    }
)

# --- Block 4: Configure and Run ---
LSTM_CONFIG = {
    "experiment_name": "UniLSTM_p2_exp01_BLVF",
    "model_type": "LSTM",
    "model_config": {
        "hidden_size": 64,
        "num_layers": 2,
        "dropout": 0.35,
        "bidirectional": True,
    },
    "data_prefix": "phas2_data",  # <-- FIX 2: Match the filename_prefix
    "splits_file": "exp-004/exp-004rolling_origin_splits.json",
    "feature_cols": feature_cols,
    "feature_selection": feature_cols, # <-- This is correct!
    "target_col": TARGET_COL,
    "batch_size": 32,
    "num_epochs": 50,
    "learning_rate": 0.001,
    "loss_function": "Huber",  # <-- This setting needs the fix below
    "early_stopping_patience": 20,
    "max_folds": 35,
}

# This will now work after you apply the fixes below
pipeline = SolarForecastingPipeline(LSTM_CONFIG)
_, summary = pipeline.run()

•	During training, the model shows decreasing loss and MAE on the training set, which is expected as the model learns the underlying patterns. However, the validation loss and MAE do not always decrease at the same rate, indicating some fluctuation in generalization.
•	Early Stopping: Many of the folds use early stopping, indicating that the model reached a point where further training did not result in significant improvements or even caused overfitting.

the model’s performance can fluctuate depending on the fold. This suggests potential instability in model predictions for some subsets of the data.

CSI metrics show that the model consistently outperforms the NAM model.

The performance drops on validation while the training loss improves, which suggests that the model is memorizing the training data instead of generalizing well.

----

In [1]:
from src.data_preparation import load_data
PROCESSED_DATA_PATH = "data/processed/dayTime_NAM_spatial_5locations_dayahead_features_processed.csv"
df = load_data(PROCESSED_DATA_PATH, date_col="measurement_time")

Loaded 11,561 records
Date range: 2014-01-03 14:00:00+00:00 to 2016-12-31 02:00:00+00:00
Timezone: UTC


In [2]:
df.columns

Index(['Unnamed: 0', 'ghi', 'dni', 'dhi', 'solar_zenith', 'time_gap_hours',
       'time_gap_norm', 'day_boundary_flag', 'hour_progression',
       'absolute_hour', 'GHI_cs', 'DNI_cs', 'CSI_ghi', 'CSI_dni',
       'season_flag', 'hour_sin', 'hour_cos', 'month_sin', 'month_cos',
       'issue_time', 'target_time', 'horizon', 'nam_ghi', 'nam_dni', 'nam_cc',
       'nam_target_time', '80_dwsw', 'valtime', '80_cloud_cover', '56_dwsw',
       '56_cloud_cover', '20_dwsw', '20_cloud_cover', '88_dwsw',
       '88_cloud_cover'],
      dtype='object')

In [3]:
# 2. Rolling Origin Split (More Suitable in TimeSeries) like K-Folds
from src.data_preparation import rolling_origin_evaluation,save_splits_info

rollingSplits_df_5locations = rolling_origin_evaluation(df=df, start_train = '2014-01-2',
    end_train = '2016-12-31')
save_splits_info({}, rollingSplits_df_5locations, experiment_name="exp-003")


ROLLING ORIGIN EVALUATION
Total folds: 35
Frequency: MS
Data range: 2014-01-03 to 2016-12-31

Fold Summary:
  Fold 1: Train [2014-01-03 to 2014-02-01] (297 records) → Val [2014-02-02 to 2014-02-28] (186 records)
  Fold 2: Train [2014-01-03 to 2014-03-01] (494 records) → Val [2014-03-02 to 2014-03-31] (319 records)
  Fold 3: Train [2014-01-03 to 2014-04-01] (835 records) → Val [2014-04-02 to 2014-04-30] (308 records)
  Fold 4: Train [2014-01-03 to 2014-05-01] (1,165 records) → Val [2014-05-02 to 2014-05-31] (319 records)
  Fold 5: Train [2014-01-03 to 2014-06-01] (1,506 records) → Val [2014-06-02 to 2014-06-30] (308 records)
  Fold 6: Train [2014-01-03 to 2014-07-01] (1,836 records) → Val [2014-07-02 to 2014-07-31] (319 records)
  Fold 7: Train [2014-01-03 to 2014-08-01] (2,177 records) → Val [2014-08-02 to 2014-08-31] (319 records)
  Fold 8: Train [2014-01-03 to 2014-09-01] (2,518 records) → Val [2014-09-02 to 2014-09-30] (308 records)
  Fold 9: Train [2014-01-03 to 2014-10-01] (2,848

In [4]:
import pandas as pd
import numpy as np
df_noIndex = df.reset_index()
df_noIndex['nam_target_time'] = pd.to_datetime(df_noIndex['nam_target_time'])
df_noIndex['measurement_time'] = pd.to_datetime(df_noIndex['measurement_time'])
df_noIndex['time_diff'] = (df_noIndex['nam_target_time'] - df_noIndex['measurement_time']).dt.total_seconds() / 3600  # Calculate the time difference in hours
    
# Check if the time difference is exactly 1 hour
is_nam_lag_correct = np.allclose(df_noIndex['time_diff'], 1)  # All time differences should be 1 hour
print(f"Does NAM records have a 1-hour lag ahead from measurements? {'Yes' if is_nam_lag_correct else 'No'}")

Does NAM records have a 1-hour lag ahead from measurements? Yes


In [5]:
# 3. Check if there are zero values in any of the features
zero_values_in_features = (df[['ghi', 'dni', 'solar_zenith', 'GHI_cs', 'DNI_cs', 'CSI_ghi', 
                                   'CSI_dni',
                                   'nam_ghi', 'nam_dni', 'nam_cc']] == 0).sum().sum()
print(f"Do we have NaN values in features? {'Yes' if zero_values_in_features > 0 else 'No'}")


features_to_check = ['ghi', 'dni', 'solar_zenith', 'GHI_cs', 'DNI_cs', 'CSI_ghi', 
                                   'CSI_dni',
                                   'nam_ghi', 'nam_dni', 'nam_cc']

# 2. Calculate the number of NaN values for each feature
# We use .isna() instead of == 0
nans_per_feature = df[features_to_check].isna().sum()

# 3. Filter to get only features that actually have NaN values
features_with_nans = nans_per_feature[nans_per_feature > 0]

# 4. Report the findings for which features have NaNs
if features_with_nans.empty:
    print("No NaN (missing) values found in any of the specified features.")
else:
    print("--- Features With NaN Values ---")
    print("The following features have NaN values, with the total count for each:")
    # Sort for clearer output
    print(features_with_nans.sort_values(ascending=False))
    print("\n" + "="*40 + "\n")

    # 5. Analyze the distribution of hours for rows containing NaNs
    print("--- Distribution of NaN-Value Records by Hour ---")
    
    # Create a boolean mask for rows that contain *at least one* NaN
    # in the specified columns
    rows_with_any_nan = df[features_to_check].isna().any(axis=1)
    
    if rows_with_any_nan.sum() > 0:
        # Get the index for these rows
        nan_rows_index = df.index[rows_with_any_nan]
        
        # Extract the hour from the DatetimeIndex and get the value counts
        hour_distribution = nan_rows_index.hour.value_counts().sort_index()
        
        print("Distribution of records (rows) containing at least one NaN, by hour:")
        print(hour_distribution)
        
        # Optional: Print total number of affected rows
        print(f"\nTotal number of rows with at least one NaN: {rows_with_any_nan.sum()}")
    else:
        # This case shouldn't be hit if features_with_nans was not empty,
        # but it's good practice to include.
        print("No rows found with NaN values (this is unexpected, check logic).")

Do we have NaN values in features? Yes
--- Features With NaN Values ---
The following features have NaN values, with the total count for each:
CSI_dni    788
CSI_ghi    639
dtype: int64


--- Distribution of NaN-Value Records by Hour ---
Distribution of records (rows) containing at least one NaN, by hour:
measurement_time
2     449
14    313
15     26
Name: count, dtype: int64

Total number of rows with at least one NaN: 788


In [6]:
df_interpolated = df.interpolate(method='linear')

/tmp/ipykernel_1694326/931910069.py:1: FutureWarning: DataFrame.interpolate with object dtype is deprecated and will raise in a future version. Call obj.infer_objects(copy=False) before interpolating instead.
  df_interpolated = df.interpolate(method='linear')


In [7]:
# 3. Check if there are zero values in any of the features
zero_values_in_features = (df_interpolated[['ghi', 'dni', 'solar_zenith', 'GHI_cs', 'DNI_cs', 'CSI_ghi', 
                                   'CSI_dni',
                                   'nam_ghi', 'nam_dni', 'nam_cc']] == 0).sum().sum()
print(f"Do we have NaN values in features? {'Yes' if zero_values_in_features > 0 else 'No'}")


features_to_check = ['ghi', 'dni', 'solar_zenith', 'GHI_cs', 'DNI_cs', 'CSI_ghi', 
                                   'CSI_dni',
                                   'nam_ghi', 'nam_dni', 'nam_cc']

# 2. Calculate the number of NaN values for each feature
# We use .isna() instead of == 0
nans_per_feature = df_interpolated[features_to_check].isna().sum()

# 3. Filter to get only features that actually have NaN values
features_with_nans = nans_per_feature[nans_per_feature > 0]

# 4. Report the findings for which features have NaNs
if features_with_nans.empty:
    print("No NaN (missing) values found in any of the specified features.")
else:
    print("--- Features With NaN Values ---")
    print("The following features have NaN values, with the total count for each:")
    # Sort for clearer output
    print(features_with_nans.sort_values(ascending=False))
    print("\n" + "="*40 + "\n")

    # 5. Analyze the distribution of hours for rows containing NaNs
    print("--- Distribution of NaN-Value Records by Hour ---")
    
    # Create a boolean mask for rows that contain *at least one* NaN
    # in the specified columns
    rows_with_any_nan = df_interpolated[features_to_check].isna().any(axis=1)
    
    if rows_with_any_nan.sum() > 0:
        # Get the index for these rows
        nan_rows_index = df_interpolated.index[rows_with_any_nan]
        
        # Extract the hour from the DatetimeIndex and get the value counts
        hour_distribution = nan_rows_index.hour.value_counts().sort_index()
        
        print("Distribution of records (rows) containing at least one NaN, by hour:")
        print(hour_distribution)
        
        # Optional: Print total number of affected rows
        print(f"\nTotal number of rows with at least one NaN: {rows_with_any_nan.sum()}")
    else:
        # This case shouldn't be hit if features_with_nans was not empty,
        # but it's good practice to include.
        print("No rows found with NaN values (this is unexpected, check logic).")

Do we have NaN values in features? Yes
--- Features With NaN Values ---
The following features have NaN values, with the total count for each:
CSI_ghi    1
CSI_dni    1
dtype: int64


--- Distribution of NaN-Value Records by Hour ---
Distribution of records (rows) containing at least one NaN, by hour:
measurement_time
14    1
Name: count, dtype: int64

Total number of rows with at least one NaN: 1


In [8]:
df_interpolated.dropna(inplace=True)

In [9]:
df_interpolated.columns

Index(['Unnamed: 0', 'ghi', 'dni', 'dhi', 'solar_zenith', 'time_gap_hours',
       'time_gap_norm', 'day_boundary_flag', 'hour_progression',
       'absolute_hour', 'GHI_cs', 'DNI_cs', 'CSI_ghi', 'CSI_dni',
       'season_flag', 'hour_sin', 'hour_cos', 'month_sin', 'month_cos',
       'issue_time', 'target_time', 'horizon', 'nam_ghi', 'nam_dni', 'nam_cc',
       'nam_target_time', '80_dwsw', 'valtime', '80_cloud_cover', '56_dwsw',
       '56_cloud_cover', '20_dwsw', '20_cloud_cover', '88_dwsw',
       '88_cloud_cover'],
      dtype='object')

In [10]:

df_phase3 = df_interpolated[['CSI_ghi', 'hour_sin', 'hour_cos', 'month_sin', 'month_cos',
    'nam_ghi', 'nam_dni', 'nam_cc',
        '80_dwsw', '80_cloud_cover', '56_dwsw',
       '56_cloud_cover', '20_dwsw', '20_cloud_cover', '88_dwsw',
       '88_cloud_cover']]


In [11]:
df_phase3.head()

CSI_ghi  hour_sin      hour_cos  month_sin  \
measurement_time                                                         
2014-01-03 15:00:00+00:00  1.200000 -0.707107 -7.071068e-01        0.5   
2014-01-03 16:00:00+00:00  1.200000 -0.866025 -5.000000e-01        0.5   
2014-01-03 17:00:00+00:00  1.200000 -0.965926 -2.588190e-01        0.5   
2014-01-03 18:00:00+00:00  1.200000 -1.000000 -1.836970e-16        0.5   
2014-01-03 19:00:00+00:00  1.126874 -0.965926  2.588190e-01        0.5   

                           month_cos  nam_ghi     nam_dni  nam_cc  80_dwsw  \
measurement_time                                                             
2014-01-03 15:00:00+00:00   0.866025   99.625  497.069277     2.0   99.625   
2014-01-03 16:00:00+00:00   0.866025  274.750  760.162689     0.0  274.750   
2014-01-03 17:00:00+00:00   0.866025  416.375  871.248807     2.0  416.375   
2014-01-03 18:00:00+00:00   0.866025  510.875  919.980698     8.0  510.875   
2014-01-03 19:00:00+00:00   0.866025  547.375  935.728503    20.0  547.375   

                           80_cloud_cover  56_dwsw  56_cloud_cover  20_dwsw  \
measurement_time                                                              
2014-01-03 15:00:00+00:00             2.0   97.375             2.0   95.000   
2014-01-03 16:00:00+00:00             0.0  273.375             0.0  274.125   
2014-01-03 17:00:00+00:00             2.0  411.500             0.0  417.500   
2014-01-03 18:00:00+00:00             8.0  508.125             4.0  509.875   
2014-01-03 19:00:00+00:00            20.0  545.375            18.0  546.750   

                           20_cloud_cover  88_dwsw  88_cloud_cover  
measurement_time                                                    
2014-01-03 15:00:00+00:00             0.0  102.000             6.0  
2014-01-03 16:00:00+00:00             0.0  279.000             2.0  
2014-01-03 17:00:00+00:00             4.0  410.375             0.0  
2014-01-03 18:00:00+00:00            12.0  508.750             2.0  
2014-01-03 19:00:00+00:00            22.0  546.000            16.0

In [12]:
from src.preprocessing import KBinConfig, process_splits_to_kbins, build_model_arrays, to_fixedgrid_multiindex
import pandas as pd
from src.utils import DataManager
from src.pipeline import SolarForecastingPipeline

# --- Block 1: Define variables ---
USE_KBINS = False
k_bins = 60
TARGET_COL = "CSI_ghi"  # <-- FIX 1: Define Target FIRST
history_days = 7
horizon_days = 1

# This list now correctly excludes the target
feature_cols = [c for c in df_phase3.columns.tolist() if c != TARGET_COL]

# --- Block 2: Build model arrays (Your code was correct here) ---
if USE_KBINS:
        print("\n--- Step 2: Converting full dataset to K-Bins format ---")
        # ... (rest of your K-Bins code) ...
else:
    print("\n--- Step 2: Building model arrays (X, Y) ---")
    fixed_df = to_fixedgrid_multiindex(df_phase3, timestamp_col="measurement_time", expected_T=None)  # or set T
    
    # You correctly passed the filtered 'feature_cols' here, fixing the data leak!
    ph3_X, ph3_Y, ph3_labels_list = build_model_arrays(
        fixed_df,
        feature_cols=feature_cols,  # <-- This is correct!
        target_col=TARGET_COL,
        history_days=history_days,
        horizon_days=horizon_days,
    )

# --- Block 3: Save arrays ---
data_manager = DataManager()
data_manager.save_arrays(
    ph3_X, ph3_Y,
    pd.DataFrame(index=pd.to_datetime(ph3_labels_list, utc=True)),
    filename_prefix='phas3_data',  # <-- You are saving as 'phas2_data'
    feature_cols=feature_cols,
    target_col=TARGET_COL,
    metadata={
        "input_csv": "data/processed/dayTime_NAM_spatial_5locations_dayahead_features_processed.csv",
        "timestamp_col": "measurement_time",
        "feature_set": feature_cols,
        "history_days": 7,
        "horizon_days": 1,
        "k_bins": None,
    }
)

# --- Block 4: Configure and Run ---
LSTM_CONFIG = {
    "experiment_name": "LSTM_phase3_exp02",
    "model_type": "LSTM",
    "model_config": {
        "hidden_size": 64,
        "num_layers": 2,
        "dropout": 0.25,
        "bidirectional": True,
    },
    "data_prefix": "phas3_data",  # <-- FIX 2: Match the filename_prefix
    "splits_file": "exp-003/exp-003rolling_origin_splits.json",
    "feature_cols": feature_cols,
    "feature_selection": feature_cols, # <-- This is correct!
    "target_col": TARGET_COL,
    "batch_size": 32,
    "num_epochs": 50,
    "learning_rate": 0.001,
    "loss_function": "Huber",  # <-- This setting needs the fix below
    "early_stopping_patience": 20,
    "max_folds": 3,
}

# This will now work after you apply the fixes below
pipeline = SolarForecastingPipeline(LSTM_CONFIG)
phas3_fold_results, summary = pipeline.run()

# To get the model from the LAST fold:
phas3_last_model = phas3_fold_results[-1]['model']

INFO:src.utils:Saved arrays to data/phas3_data_*.npy
INFO:src.utils:X shape: (1056, 7, 11, 15), Y shape: (1056, 1, 11)



--- Step 2: Building model arrays (X, Y) ---


INFO:src.pipeline:Loading data...
INFO:src.utils:Loaded arrays from data/phas3_data_*.npy
INFO:src.utils:X shape: (1056, 7, 11, 15), Y shape: (1056, 1, 11)
INFO:src.pipeline:Building FIXED-GRID reference frame (K_bins is None)...
INFO:src.utils:Loaded 35 folds from exp-003/exp-003rolling_origin_splits.json
INFO:src.pipeline:
=== Running Fold 1 ===
INFO:src.pipeline:Train samples: 21, Val samples: 18
INFO:src.pipeline:Model parameters: 251,531
INFO:src.engine:Using HuberLoss (Smooth L1 Loss)
INFO:src.engine:Epoch [10/50] - Train Loss: 0.048390, Val Loss: 0.080466, Train MAE: 0.222599, Val MAE: 0.313158, LR: 0.001000
INFO:src.engine:Epoch [20/50] - Train Loss: 0.032025, Val Loss: 0.061464, Train MAE: 0.179580, Val MAE: 0.289278, LR: 0.001000
INFO:src.engine:Epoch [30/50] - Train Loss: 0.026566, Val Loss: 0.061203, Train MAE: 0.152212, Val MAE: 0.273124, LR: 0.001000
INFO:src.engine:Epoch [40/50] - Train Loss: 0.021560, Val Loss: 0.089690, Train MAE: 0.133791, Val MAE: 0.317289, LR: 0.001

In [12]:
from src.preprocessing import KBinConfig, process_splits_to_kbins, build_model_arrays, to_fixedgrid_multiindex
import pandas as pd
from src.utils import DataManager
from src.pipeline import SolarForecastingPipeline


df_phase4 = df_interpolated[['CSI_ghi', 'hour_sin', 'hour_cos', 'month_sin', 'month_cos',
    'nam_ghi', 'nam_cc']]


# --- Block 1: Define variables ---
USE_KBINS = False
k_bins = 60
TARGET_COL = "CSI_ghi"  # <-- FIX 1: Define Target FIRST
history_days = 7
horizon_days = 1

# This list now correctly excludes the target
feature_cols = [c for c in df_phase4.columns.tolist() if c != TARGET_COL]

# --- Block 2: Build model arrays (Your code was correct here) ---
if USE_KBINS:
        print("\n--- Step 2: Converting full dataset to K-Bins format ---")
        # ... (rest of your K-Bins code) ...
else:
    print("\n--- Step 2: Building model arrays (X, Y) ---")
    fixed_df = to_fixedgrid_multiindex(df_phase4, timestamp_col="measurement_time", expected_T=None)  # or set T
    
    # You correctly passed the filtered 'feature_cols' here, fixing the data leak!
    ph4_X, ph4_Y, ph4_labels_list = build_model_arrays(
        fixed_df,
        feature_cols=feature_cols,  # <-- This is correct!
        target_col=TARGET_COL,
        history_days=history_days,
        horizon_days=horizon_days,
    )

# --- Block 3: Save arrays ---
data_manager = DataManager()
data_manager.save_arrays(
    ph4_X, ph4_Y,
    pd.DataFrame(index=pd.to_datetime(ph4_labels_list, utc=True)),
    filename_prefix='phas4_data',  # <-- You are saving as 'phas2_data'
    feature_cols=feature_cols,
    target_col=TARGET_COL,
    metadata={
        "input_csv": "data/processed/dayTime_NAM_spatial_5locations_dayahead_features_processed.csv",
        "timestamp_col": "measurement_time",
        "feature_set": feature_cols,
        "history_days": 7,
        "horizon_days": 1,
        "k_bins": None,
    }
)

# --- Block 4: Configure and Run ---
LSTM_CONFIG = {
    "experiment_name": "LSTM_phase4_exp02",
    "model_type": "LSTM",
    "model_config": {
        "hidden_size": 64,
        "num_layers": 2,
        "dropout": 0.25,
        "bidirectional": True,
    },
    "data_prefix": "phas4_data",  # <-- FIX 2: Match the filename_prefix
    "splits_file": "exp-003/exp-003rolling_origin_splits.json",
    "feature_cols": feature_cols,
    "feature_selection": feature_cols, # <-- This is correct!
    "target_col": TARGET_COL,
    "batch_size": 32,
    "num_epochs": 50,
    "learning_rate": 0.001,
    "loss_function": "Huber",  # <-- This setting needs the fix below
    "early_stopping_patience": 20,
    "max_folds": 35,
}

# This will now work after you apply the fixes below
pipeline = SolarForecastingPipeline(LSTM_CONFIG)
fold_results, summary = pipeline.run()

# To get the model from the LAST fold:
last_model = fold_results[-1]['model']

INFO:src.utils:Saved arrays to data/phas4_data_*.npy
INFO:src.utils:X shape: (1056, 7, 11, 6), Y shape: (1056, 1, 11)
INFO:src.pipeline:Loading data...
INFO:src.utils:Loaded arrays from data/phas4_data_*.npy
INFO:src.utils:X shape: (1056, 7, 11, 6), Y shape: (1056, 1, 11)


INFO:src.utils:Loaded 35 folds from exp-003/exp-003rolling_origin_splits.json
INFO:src.pipeline:
=== Running Fold 1 ===
INFO:src.pipeline:Train samples: 21, Val samples: 18
INFO:src.pipeline:Model parameters: 245,195
INFO:src.engine:Using HuberLoss (Smooth L1 Loss)



--- Step 2: Building model arrays (X, Y) ---


INFO:src.engine:Epoch [10/50] - Train Loss: 0.051697, Val Loss: 0.087358, Train MAE: 0.227353, Val MAE: 0.320367, LR: 0.001000
INFO:src.engine:Epoch [20/50] - Train Loss: 0.044961, Val Loss: 0.076549, Train MAE: 0.212394, Val MAE: 0.301476, LR: 0.001000
INFO:src.engine:Epoch [30/50] - Train Loss: 0.024929, Val Loss: 0.077833, Train MAE: 0.152070, Val MAE: 0.294396, LR: 0.001000
INFO:src.engine:Epoch [40/50] - Train Loss: 0.023551, Val Loss: 0.081485, Train MAE: 0.134832, Val MAE: 0.297576, LR: 0.000500
INFO:src.engine:Early stopping at epoch 45
INFO:src.pipeline:Fold 1 - Validation Metrics:
INFO:src.pipeline:  mse: 0.161004
INFO:src.pipeline:  rmse: 0.401253
INFO:src.pipeline:  mae: 0.295937
INFO:src.pipeline:  r2: 0.027275
INFO:src.pipeline:  mape: 78.607788
INFO:src.pipeline:
=== Running Fold 2 ===
INFO:src.pipeline:Train samples: 40, Val samples: 29
INFO:src.pipeline:Model parameters: 245,195
INFO:src.engine:Using HuberLoss (Smooth L1 Loss)
INFO:src.engine:Epoch [10/50] - Train Loss

In [13]:
from src.preprocessing import KBinConfig, process_splits_to_kbins, build_model_arrays, to_fixedgrid_multiindex
import pandas as pd
from src.utils import DataManager
from src.pipeline import SolarForecastingPipeline


df_phase5 = df_interpolated[['CSI_ghi', 'hour_sin', 'hour_cos', 'month_sin', 'month_cos',
     'nam_cc',
        '80_cloud_cover',
       '56_cloud_cover', '20_cloud_cover',
       '88_cloud_cover']]


# --- Block 1: Define variables ---
USE_KBINS = False
k_bins = 60
TARGET_COL = "CSI_ghi"  # <-- FIX 1: Define Target FIRST
history_days = 7
horizon_days = 1

# This list now correctly excludes the target
feature_cols = [c for c in df_phase5.columns.tolist() if c != TARGET_COL]

# --- Block 2: Build model arrays (Your code was correct here) ---
if USE_KBINS:
        print("\n--- Step 2: Converting full dataset to K-Bins format ---")
        # ... (rest of your K-Bins code) ...
else:
    print("\n--- Step 2: Building model arrays (X, Y) ---")
    fixed_df = to_fixedgrid_multiindex(df_phase5, timestamp_col="measurement_time", expected_T=None)  # or set T
    
    # You correctly passed the filtered 'feature_cols' here, fixing the data leak!
    ph5_X, ph5_Y, ph5_labels_list = build_model_arrays(
        fixed_df,
        feature_cols=feature_cols,  # <-- This is correct!
        target_col=TARGET_COL,
        history_days=history_days,
        horizon_days=horizon_days,
    )

# --- Block 3: Save arrays ---
data_manager = DataManager()
data_manager.save_arrays(
    ph5_X, ph5_Y,
    pd.DataFrame(index=pd.to_datetime(ph5_labels_list, utc=True)),
    filename_prefix='phas5_data',  # <-- You are saving as 'phas2_data'
    feature_cols=feature_cols,
    target_col=TARGET_COL,
    metadata={
        "input_csv": "data/processed/dayTime_NAM_spatial_5locations_dayahead_features_processed.csv",
        "timestamp_col": "measurement_time",
        "feature_set": feature_cols,
        "history_days": 7,
        "horizon_days": 1,
        "k_bins": None,
    }
)

# --- Block 4: Configure and Run ---
LSTM_CONFIG = {
    "experiment_name": "LSTM_phase5_exp02",
    "model_type": "LSTM",
    "model_config": {
        "hidden_size": 64,
        "num_layers": 2,
        "dropout": 0.25,
        "bidirectional": True,
    },
    "data_prefix": "phas5_data",  # <-- FIX 2: Match the filename_prefix
    "splits_file": "exp-003/exp-003rolling_origin_splits.json",
    "feature_cols": feature_cols,
    "feature_selection": feature_cols, # <-- This is correct!
    "target_col": TARGET_COL,
    "batch_size": 32,
    "num_epochs": 50,
    "learning_rate": 0.001,
    "loss_function": "Huber",  # <-- This setting needs the fix below
    "early_stopping_patience": 20,
    "max_folds": 35,
}

# This will now work after you apply the fixes below
pipeline = SolarForecastingPipeline(LSTM_CONFIG)
phas5_fold_results, summary = pipeline.run()

# To get the model from the LAST fold:
phas5_fold_results = phas3_fold_results[-1]['model']

INFO:src.utils:Saved arrays to data/phas5_data_*.npy
INFO:src.utils:X shape: (1056, 7, 11, 9), Y shape: (1056, 1, 11)
INFO:src.pipeline:Loading data...
INFO:src.utils:Loaded arrays from data/phas5_data_*.npy
INFO:src.utils:X shape: (1056, 7, 11, 9), Y shape: (1056, 1, 11)
INFO:src.utils:Loaded 35 folds from exp-003/exp-003rolling_origin_splits.json
INFO:src.pipeline:
=== Running Fold 1 ===
INFO:src.pipeline:Train samples: 21, Val samples: 18
INFO:src.pipeline:Model parameters: 247,307
INFO:src.engine:Using HuberLoss (Smooth L1 Loss)



--- Step 2: Building model arrays (X, Y) ---


INFO:src.engine:Epoch [10/50] - Train Loss: 0.049739, Val Loss: 0.088166, Train MAE: 0.221810, Val MAE: 0.325361, LR: 0.001000
INFO:src.engine:Epoch [20/50] - Train Loss: 0.025353, Val Loss: 0.074557, Train MAE: 0.148397, Val MAE: 0.294710, LR: 0.001000
INFO:src.engine:Epoch [30/50] - Train Loss: 0.022186, Val Loss: 0.080162, Train MAE: 0.141149, Val MAE: 0.305926, LR: 0.000500
INFO:src.engine:Early stopping at epoch 38
INFO:src.pipeline:Fold 1 - Validation Metrics:
INFO:src.pipeline:  mse: 0.167472
INFO:src.pipeline:  rmse: 0.409233
INFO:src.pipeline:  mae: 0.301700
INFO:src.pipeline:  r2: -0.011800
INFO:src.pipeline:  mape: 70.875938
INFO:src.pipeline:
=== Running Fold 2 ===
INFO:src.pipeline:Train samples: 40, Val samples: 29
INFO:src.pipeline:Model parameters: 247,307
INFO:src.engine:Using HuberLoss (Smooth L1 Loss)
INFO:src.engine:Epoch [10/50] - Train Loss: 0.055114, Val Loss: 0.061075, Train MAE: 0.258907, Val MAE: 0.291307, LR: 0.001000
INFO:src.engine:Epoch [20/50] - Train Los

Inconsistent Results Across Folds: When adding the spatial-temporal features, the model shows considerable variation in performance across the folds. Some folds perform worse (e.g., Fold 3: RMSE 0.2124 to 0.466), while others perform better (e.g., Fold 31: RMSE 0.0918). This suggests that the added features may be contributing noise or irrelevant complexity, which the model is struggling to handle uniformly across different subsets of the data.

The higher variance in metrics across folds, especially in terms of RMSE and MAE, indicates that the model is not stable with the added features. This suggests that while the added features may hold some information, their inclusion doesn’t consistently improve predictive accuracy across all training and validation splits.

In [49]:
fold_results

[{'history': {'train_loss': [0.06003684177994728,
    0.059081628918647766,
    0.05886007845401764,
    0.05694744363427162,
    0.0553533211350441,
    0.053611330687999725,
    0.05284370481967926,
    0.052881281822919846,
    0.05291888490319252,
    0.05297527462244034,
    0.04997491464018822,
    0.053134702146053314,
    0.04904967173933983,
    0.0478777177631855,
    0.047167930752038956,
    0.046798259019851685,
    0.04694787412881851,
    0.04371234029531479,
    0.044653840363025665,
    0.0406917929649353,
    0.04292237013578415,
    0.03734507039189339,
    0.03371132165193558,
    0.030639274045825005,
    0.03304049372673035,
    0.029997609555721283,
    0.02925606444478035,
    0.02697451412677765,
    0.02603142336010933,
    0.026743434369564056,
    0.024257680401206017,
    0.02271345816552639,
    0.02418750710785389,
    0.028082501143217087,
    0.02268541231751442,
    0.018612494692206383,
    0.022291239351034164,
    0.023941392078995705,
    0.0206559

In [ ]:
from src.preprocessing import KBinConfig, process_splits_to_kbins, build_model_arrays, to_fixedgrid_multiindex
import pandas as pd
from src.utils import DataManager
from src.pipeline import SolarForecastingPipeline


df_phase5 = df_interpolated[['CSI_ghi', 'hour_sin', 'hour_cos', 'month_sin', 'month_cos',
     'nam_cc',
        '80_cloud_cover',
       '56_cloud_cover', '20_cloud_cover',
       '88_cloud_cover']]


# --- Block 1: Define variables ---
USE_KBINS = False
k_bins = 60
TARGET_COL = "CSI_ghi"  # <-- FIX 1: Define Target FIRST
history_days = 7
horizon_days = 1

# This list now correctly excludes the target
feature_cols = [c for c in df_phase5.columns.tolist() if c != TARGET_COL]

# --- Block 2: Build model arrays (Your code was correct here) ---
if USE_KBINS:
        print("\n--- Step 2: Converting full dataset to K-Bins format ---")
        # ... (rest of your K-Bins code) ...
else:
    print("\n--- Step 2: Building model arrays (X, Y) ---")
    fixed_df = to_fixedgrid_multiindex(df_phase5, timestamp_col="measurement_time", expected_T=None)  # or set T
    
    # You correctly passed the filtered 'feature_cols' here, fixing the data leak!
    ph5_X, ph5_Y, ph5_labels_list = build_model_arrays(
        fixed_df,
        feature_cols=feature_cols,  # <-- This is correct!
        target_col=TARGET_COL,
        history_days=history_days,
        horizon_days=horizon_days,
    )

# --- Block 3: Save arrays ---
data_manager = DataManager()
data_manager.save_arrays(
    ph5_X, ph5_Y,
    pd.DataFrame(index=pd.to_datetime(ph5_labels_list, utc=True)),
    filename_prefix='phas5_data',  # <-- You are saving as 'phas2_data'
    feature_cols=feature_cols,
    target_col=TARGET_COL,
    metadata={
        "input_csv": "data/processed/dayTime_NAM_spatial_5locations_dayahead_features_processed.csv",
        "timestamp_col": "measurement_time",
        "feature_set": feature_cols,
        "history_days": 7,
        "horizon_days": 1,
        "k_bins": None,
    }
)

# --- Block 4: Configure and Run ---
LSTM_CONFIG = {
    "experiment_name": "LSTM_UniDirectional_phase5_exp02",
    "model_type": "LSTM",
    "model_config": {
        "hidden_size": 64,
        "num_layers": 2,
        "dropout": 0.25,
        "bidirectional": False,
    },
    "data_prefix": "phas5_data",  # <-- FIX 2: Match the filename_prefix
    "splits_file": "exp-003/exp-003rolling_origin_splits.json",
    "feature_cols": feature_cols,
    "feature_selection": feature_cols, # <-- This is correct!
    "target_col": TARGET_COL,
    "batch_size": 32,
    "num_epochs": 50,
    "learning_rate": 0.001,
    "loss_function": "Huber",  # <-- This setting needs the fix below
    "early_stopping_patience": 20,
    "max_folds": 35,
}

# This will now work after you apply the fixes below
pipeline = SolarForecastingPipeline(LSTM_CONFIG)
phas5_LSTM_UniDir_fold_results, summary = pipeline.run()

# To get the model from the LAST fold:
phas5_LSTM_UniDir_fold_results_model = phas5_LSTM_UniDir_fold_results[-1]['model']

INFO:src.utils:Saved arrays to data/phas5_data_*.npy
INFO:src.utils:X shape: (1056, 7, 11, 9), Y shape: (1056, 1, 11)
INFO:src.pipeline:Loading data...
INFO:src.utils:Loaded arrays from data/phas5_data_*.npy
INFO:src.utils:X shape: (1056, 7, 11, 9), Y shape: (1056, 1, 11)



--- Step 2: Building model arrays (X, Y) ---


INFO:src.utils:Loaded 35 folds from exp-003/exp-003rolling_origin_splits.json
INFO:src.pipeline:
=== Running Fold 1 ===
INFO:src.pipeline:Train samples: 21, Val samples: 18
INFO:src.pipeline:Model parameters: 94,475
INFO:src.engine:Using HuberLoss (Smooth L1 Loss)
INFO:src.engine:Epoch [10/50] - Train Loss: 0.056521, Val Loss: 0.097582, Train MAE: 0.241174, Val MAE: 0.341793, LR: 0.001000
INFO:src.engine:Epoch [20/50] - Train Loss: 0.051135, Val Loss: 0.089080, Train MAE: 0.223228, Val MAE: 0.332661, LR: 0.001000
INFO:src.engine:Epoch [30/50] - Train Loss: 0.044316, Val Loss: 0.078961, Train MAE: 0.192104, Val MAE: 0.334169, LR: 0.001000
INFO:src.engine:Epoch [40/50] - Train Loss: 0.027990, Val Loss: 0.184373, Train MAE: 0.153841, Val MAE: 0.511356, LR: 0.001000
INFO:src.engine:Epoch [50/50] - Train Loss: 0.024508, Val Loss: 0.144790, Train MAE: 0.150110, Val MAE: 0.462067, LR: 0.000500
INFO:src.pipeline:Fold 1 - Validation Metrics:
INFO:src.pipeline:  mse: 0.289604
INFO:src.pipeline: 

Interactive plot saved successfully to: solar_irradiance_fold35.html
Open this file in any web browser to view the interactive visualization.
